### DISCUSSION TITLE
This script should only be loaded in a browser extension.

### DISCUSSION THREAD

**Discussion by xunziheng** [wxt-dev/wxt#700] [Q&A] [answered]
when I write browser.xxx in content script ui within main world.

it doesn't work. so why, maybe it's not allowed to use browser api in main world content script?

```js
browser.runtime.sendMessage(message).then((response) => {
  console.log('sw response: ', response);
});
```

<img width="544" alt="image" src="https://github.com/wxt-dev/wxt/assets/35691153/1234e4a9-cbc8-4e76-bdda-5dc1bb397b2f">

**Comment by xunziheng** [comment]
should I use `window.postMessage` to messaging between `main world` and `isolated world`, then use `browser.runtime.sendMessage` to messaging between `isolated world content script` and `service work`?

**Reply by aklinker1** [reply, marked as answer]
Main world content scripts don't have access to the web extension APIs, other than external messaging [if you set that up](https://developer.chrome.com/docs/extensions/develop/concepts/messaging#external-webpage).

`window.postMessage` will work as well.
